layerindex/tools/import_layers.py

```diff
@@ -0,0 +1,494 @@
+                                       custom_field_cb=recipe_field_handler,
+                                       key_fields=['pn'])
+
+                if machines_url:
+                    import_child_items(layerbranch,
+                                       Machine,
+                                       url=machines_url,
+                                       parent_orig_id=layerbranchjs['id'],
+                                       exclude_fields=['id', 'layerbranch', 'updated'],
+                                       key_fields=['name'])
+
+                if distros_url:
+                    import_child_items(layerbranch,
+                                       Distro,
+                                       url=distros_url,
+                                       parent_orig_id=layerbranchjs['id'],
+                                       exclude_fields=['id', 'layerbranch', 'updated'],
+                                       key_fields=['name'])
+
+                # The models below don't have an "updated" field at present, but it does
+                # no harm to leave it as excluded in case it does get added in the future
+
+                if classes_url:
+                    import_child_items(layerbranch,
+                                       BBClass,
+                                       url=classes_url,
+                                       parent_orig_id=layerbranchjs['id'],
+                                       exclude_fields=['id', 'layerbranch', 'updated'],
+                                       key_fields=['name'])
+
+                if appends_url:
+                    import_child_items(layerbranch,
+                                       BBAppend,
+                                       url=appends_url,
+                                       parent_orig_id=layerbranchjs['id'],
+                                       exclude_fields=['id', 'layerbranch', 'updated'],
+                                       key_fields=['filename'])
+
+                if incfiles_url:
+                    import_child_items(layerbranch,
+                                       IncFile,
+                                       url=incfiles_url,
+                                       parent_orig_id=layerbranchjs['id'],
+                                       exclude_fields=['id', 'layerbranch', 'updated'],
+                                       key_fields=['path'])
+
+            for idv in existing_layerbranches:
+                layerbranch = LayerBranch.objects.get(id=idv)
+                if layer_re is None or layer_re.match(layerbranch.layer.name):
+                    logger.debug('Deleting layerbranch %s' % layerbranch)
+                    layerbranch.delete()
+
+            # Get layer dependencies
+            logger.info('Importing layer dependencies')
+            rq = urllib.request.Request(layerdeps_url)
+            data = urllib.request.urlopen(rq).read()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *urlopen*. The resource is allocated by call *request.urlopen*. Execution paths that do not contain closure statements were detected. To prevent this resource leak, close the object returned by *urlopen()* in a try-finally block or declare it using a `with` statement.

**More info**
[View details about the `with` statement in the Python developer&#39;s guide](https://www.python.org/dev/peps/pep-0343/) (external link).